### PR TITLE
Add BearSSL documentation

### LINKED ISSUE #1743: Add documentation for WiFiClientSecure (including usage of client side certificates)
Hi,

I am using Mqtt with esp8266. I am wondering does esp8266 support clients certificate for authentication? 

Thanks for the advice.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31717132-add-documentation-for-wificlientsecure-including-usage-of-client-side-certificates?utm_campaign=plugin&utm_content=tracker%2F14245935&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F14245935&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by earlephilhower**
Document the BearSSL::WiFiClientSecure, ::WiFiServerSecure, and the
helper classes required to use them.

Fixes #1743 and other SSL documentation issues.

### CODE REVIEW

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+TLS Sessions
+~~~~~~~~~~~~
+
+TLS supports the notion of a session (completely independent and different from HTTP sessions) which allow clients to reconnect to a server without having to renegotiate encryption settings or validate X509 certificates.  This can save significant time (3-4 seconds in the case of EC keys) and can help save power by allowing the ESP8266 to sleep for a long time, reconnect and transmit some samples using the SSL session, and then jump back to sleep quicker.
+
+`BearSSLSession` is an opaque class.  Use the `BearSSL::WiFiClientSecure.setSession(&BearSSLSession)` method to apply it before the first `BearSSL::WiFiClientSecure.connect()` and it will be updated with session parameters during the operation of the connection.  After the connection had had `.close()` called on it, serialize the `BearSSLSession` object to stable store (EEPROM, RTC RAM, etc) and restore it before trying to resonnect.  See the `BearSSL_Sessions` example for a detailed example.
```
[the remark this answers is not in the document]
Comment (devyte) [reply doc/esp8266wifi/bearssl-client-secure-class.rst]: stable store => stable storage?

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+
+X.509 Certificate(s)
+~~~~~~~~~~~~~~~~~~~~
+
+X509 certificates are used to identify peers in TLS connections.  Normally only the server identifies itself, but the client can also supply an X509 certificate if desired (this is often done in MQTT applications).  The certificate contains many fields, but the most interesting in our applications are the name, the public key, and potentially a chain of signing that leads back to a trusted authority (like a global internet CA or a company-wide private certificate authority).
+
+Any call that takes an X509 certificate can also take a list of X509 certificates, so there is no special `X509` class, simply `BearSSLX509List` (which may only contain a sincle certificate).
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: sincle => single

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+Generating a certificate to be used to validate using the constructor
+
+.. code:: cpp
+
+    BearSSLX509List(const char *pemX509);
+    ...or...
+    BearSSLX509List(const uint8_t *derCert, size_t derLen);
+
+If you need to add additional certificates (unlikely in normal operation, the `::append()` operation can be used.
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: missing closing ), maybe before the comma?

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+setKnownKey(const BearSSLPublicKey *pk)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Assume the server is using the specific public key.  This does not verify the identity of the server or the X509 certificate it sends, it simply assumes that its public key is the one given.  If the server updates its public key at a later point then connections will fail.
+
+setFingerprint(const uint8_t fp[20]) / setFingerprint(const char *fpStr)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Verify the SHA1 fingerprint of the certificate returned matches this one.  If the server certificate changes, it will fail.  If an array of 20 bytes are sent in, it is assumed they are the binary SHA1 values.  If a `char*` string is passed in, it is parsed as a series of human-readable hex values separated by spaces or colons (i.e. `setFingerprint("00:01:02:03:...:1f");`)
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: i.e. => e.g.

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+setX509Time(time_t now)
+^^^^^^^^^^^^^^^^^^^^^^^
+
+For `setTrustAnchors` and `CertStore` , the current time (set via SNTP) is used to verify the certificate against the list, so SNTP must be enabled and functioning before the connection is attempted.  If you cannot use SNTP for some reason, you can manually set the "present time" that BearSSL will use to validate a certificate with this call where `now` is standard UNIX time.
+
+Client Certificates (Proving I'm who I say I am to the server)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+TLS servers can request that a client identify themselves with an X509 certificate signed by a trust anchor it honors (i.e. a global TA or a private CA).  This is commonly done for applicaitons like MQTT.  By default the client doesn't sent a certificate, and in cases where a certificate is required the server will disconnect and no connection will be possible.
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: sent => send

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+setClientRSACert / setClientECCert
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sets a client certificate to send to a TLS server that requests one.  It should be called before `connect()` to add a certificate to the client in case the server requests is.  Note that certificates include both a certificate and a private key.  Both should be provided to you by your certificate generator.  Elliptic Curve (EC) keys require additional information, as shown in the prototype.
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: requests is => requests it

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Use one of these calls **before** connection to determine if a specific fragment length is supported (len must be a power of two from 512 to 4096, per the TLC specification).  This does **not** initiate a SSL connection, it simply opens a TCP port and performs a trial handshake to check support.
+
+setBufferSizes(int recv, int xmit)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once you have verified (or know beforehand) that MFLN is supported you can use this call to set the size of memory buffers allocated by the connection object.  This must be called **before** `connect()` or it will be ignored.
+
+In certain applications where the TLS server does not support MFLN (not many do as of this writing as it is relatively new to OpenSSL) but you control both the ESP8266 and the server to which it is communicating, you may still be able to `setBufferSizes()` smaller if you guarantee no chunk of data will overflow those buffers.
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: OpenSSL) but you => OpenSSL), but you (add comma)

File: doc/esp8266wifi/bearssl-client-secure-class.rst [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you are connecting to a server repeatedly in a fixed time period (usually 30 or 60 minuntes, but normally configurable at the server), a TLS session can be used to cache crypto settings and speed up connections significantly.
+
+Errors
+~~~~~~
+
+BearSSL can fail in many more unique and interesting ways then the deprecated axTLS.  Use these calls to get more information when something fails.  
+
+getLastSSLError(char *dest = NULL, size_t len = 0)
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-client-secure-class.rst]: Is the user supposed to allocate some buffer?
Or is the arg supposed to be a pointer to the base of a static-sized buffer?
Or is the pointer set to an internal static string? If so, should the args be const?

File: doc/esp8266wifi/bearssl-server-secure-class.rst [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+BearSSL Secure Server Class
+---------------------------
+
+Implements a TLS encrypted server with optional client certificate validation.  See `Server Class <server-class.rst>`__ for general information and `BearSSL Secure Client Class <bearssl-secure-client-class.rst>`__ for basic server and BearSSL concepts.
+
+setBufferSizes(int recv, int xmit)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to the `BearSSL::WiFiClientSecure` methos, sets the receive and transmit buffer sizes.  Note that servers cannot request a buffer size from the client, so if these are shrunk and the client tries to send a chunk larger than the receive buffer, it will always fail.  This must be called before the server is
```
Comment (devyte) [inline doc/esp8266wifi/bearssl-server-secure-class.rst]: methos => method

**Comment by earlephilhower** [comment]
Thanks, @devyte, for the read-through.  Updated with your grammar and spelling (plus MSFT Word spellcheck) and cleaned up the `getLastError()` bit to be more specific that user allocates and manages the string buffer optionally passed in.

**Pull request merged**